src/Security/MemberAuthenticator/LoginHandler.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+namespace SilverStripe\Security\MemberAuthenticator;
+
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\Session;
+use SilverStripe\Control\RequestHandler;
+use SilverStripe\ORM\ValidationResult;
+use SilverStripe\Security\MemberAuthenticator\Authenticator;
+use SilverStripe\Security\Security;
+use SilverStripe\Security\Member;
+
+/**
+ * Handle login requests from MemberLoginForm
+ */
+class LoginHandler extends RequestHandler
+{
+    protected $authenticator;
+
+    private static $url_handlers = [
+        '' => 'login',
+    ];
+
+    /**
+     * Since the logout and dologin actions may be conditionally removed, it's necessary to ensure these
+     * remain valid actions regardless of the member login state.
+     *
+     * @var array
+     * @config
+     */
+    private static $allowed_actions = [
+        'login',
+        'LoginForm',
+        'dologin',
+        'logout',
+    ];
+
+    private $link = null;
+
+    /**
+     * @param string $link The URL to recreate this request handler
+     * @param Authenticator $authenticator The
+     */
+    public function __construct($link, Authenticator $authenticator)
+    {
+        $this->link = $link;
+        $this->authenticator = $authenticator;
+        parent::__construct($link, $this);
+    }
+
+    /**
+     * Return a link to this request handler.
+     * The link returned is supplied in the constructor
+     * @return string
+     */
+    public function link($action = null)
+    {
+        if ($action) {
+            return Controller::join_links($this->link, $action);
+        } else {
+            return $this->link;
+        }
+    }
+
+    /**
+     * URL handler for the log-in screen
+     */
+    public function login()
+    {
+        return [
+            'Form' => $this->loginForm(),
+        ];
+    }
+
+    /**
+     * Return the MemberLoginForm form
+     */
+    public function loginForm()
+    {
+        return LoginForm::create(
+            $this,
+            get_class($this->authenticator),
+            'LoginForm'
+        );
+    }
+
+    /**
+     * Login form handler method
+     *
+     * This method is called when the user clicks on "Log in"
+     *
+     * @param array $data Submitted data
+     * @param LoginHandler $formHandler
+     * @return HTTPResponse
+     */
+    public function doLogin($data, $form)
+    {
+        $failureMessage = null;
+
+        // Successful login
+        if ($member = $this->checkLogin($data, $failureMessage)) {
+            $this->performLogin($member, $data);
+            return $this->redirectAfterSuccessfulLogin();
+        }
+
+        $form->sessionMessage($failureMessage, 'bad');
```

Comment:
Shouldn't this have been handled way earlier already? In checkLogin for example? I feel `doLogin` should not be the one handling the errors, at `doLogin`, the member is in the safe zone.